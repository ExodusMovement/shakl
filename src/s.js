import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
