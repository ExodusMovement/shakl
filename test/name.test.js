import React from 'react';

import { create as r } from 'react-test-renderer';

import { View } from 'react-native';

import s from '../src';

it('has proper display name', () => {
  const Foo = s.Text({ color: 'red' });
  const foo = r(<Foo />).toJSON();
  expect(Foo.displayName).toBe('styled(Text)');
  expect(foo).toMatchSnapshot();
});

it('allows providing a custom display name', () => {
  const Foo = s(View, { name: 'Foo' })({ flex: 1 });
  const foo = r(<Foo />).toJSON();
  expect(Foo.displayName).toBe('Foo');
  expect(foo).toMatchSnapshot();
});

it('falls back to Component.name if Component.displayName cannot be inferred', () => {
  const Foo = () => null;
  const Bar = s(Foo)();
  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  expect(Foo.displayName).not.toBeDefined();
  expect(Foo.name).toBe('Foo');
  expect(Bar.displayName).toBe('styled(Foo)');
  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

it('keeps custom display name when extended', () => {
  const Foo = s(View, { name: 'Foo' })({ flex: 1 });
  const Bar = Foo.extend({ flex: 1 });
  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  expect(Foo.displayName).toBe('Foo');
  expect(Bar.displayName).toBe('Foo');
  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
});

test('exposed primitives have proper display names', () => {
  const Foo = s.Text({ color: 'red' });
  const Bar = s.Touchable({ flex: 1 });
  const Baz = s.View({ flex: 1 });
  const foo = r(<Foo />).toJSON();
  const bar = r(<Bar />).toJSON();
  const baz = r(<Baz />).toJSON();
  expect(Foo.displayName).toBe('styled(Text)');
  expect(Bar.displayName).toBe('styled(Touchable)');
  expect(Baz.displayName).toBe('styled(View)');
  expect(foo).toMatchSnapshot();
  expect(bar).toMatchSnapshot();
  expect(baz).toMatchSnapshot();
});
