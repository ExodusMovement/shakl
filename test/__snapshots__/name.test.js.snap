// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows providing a custom display name 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
/>
`;

exports[`exposed primitives have proper display names 1`] = `
<Text
  style={
    {
      "color": "red",
    }
  }
/>
`;

exports[`exposed primitives have proper display names 2`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "flex": 1,
      "opacity": 1,
    }
  }
/>
`;

exports[`exposed primitives have proper display names 3`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
/>
`;

exports[`falls back to Component.name if Component.displayName cannot be inferred 1`] = `null`;

exports[`falls back to Component.name if Component.displayName cannot be inferred 2`] = `null`;

exports[`has proper display name 1`] = `
<Text
  style={
    {
      "color": "red",
    }
  }
/>
`;

exports[`keeps custom display name when extended 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
/>
`;

exports[`keeps custom display name when extended 2`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
/>
`;
